sound/soc/sof/topology.c

```diff
@@ -1428,6 +1445,53 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
 	return 0;
 }
 
+/**
+ * sof_comp_alloc - allocate and initialize buffer for a new component
+ * @swidget: pointer to struct snd_sof_widget containing extended data
+ * @ipc_size: IPC payload size that will be updated depending on valid
+ *  extended data.
+ * @index: ID of the pipeline the component belongs to
+ * @core: index of the DSP core that the component should run on
+ *
+ * Return: The pointer to the new allocated component, NULL if failed.
+ */
+static struct sof_ipc_comp *sof_comp_alloc(struct snd_sof_widget *swidget,
+					   size_t *ipc_size, int index,
+					   int core)
+{
+	struct sof_ipc_comp_ext nul_ext_data = {0};
+	struct sof_ipc_comp *comp;
+	size_t total_size = *ipc_size;
+
+	/* only non-zero ext_data is valid */
+	if (memcmp(&swidget->comp_ext, &nul_ext_data,
+		   sizeof(swidget->comp_ext)))
```

Comment:
I still think the following code is more future-proof:
```C
uint8_t nul_uuid[SOF_UUID_SIZE] = {0};
...
if (memcmp(&swidget->comp_ext.uuid, &nul_uuid, SOF_UUID_SIZE)

```

We had incredibly long discussions about extended_data_length/offset etc., so there's clearly some desire for extended data to be not just an UUID in the future. When that happens, we don't want to allocate and compare a potentially not so small array of zeroes on the kernel stack for no good reason.

Hey we could even want different components to have different lengths of extended data because why not? This would of course require more code changes but at least this `memcmp` would not have to change.

Reply:
> I still think the following code is more future-proof:
> 
> ```c
> uint8_t nul_uuid[SOF_UUID_SIZE] = {0};
> ...
> if (memcmp(&swidget->comp_ext.uuid, &nul_uuid, SOF_UUID_SIZE)
> ```
> 
> We had incredibly long discussions about extended_data_length/offset etc., so there's clearly some desire for extended data to be not just an UUID in the future. When that happens, we don't want to allocate and compare a potentially not so small array of zeroes on the kernel stack for no good reason.

Well, this is actually not very easy to be handled here. Ideally if we can check the existence of each extended data member and add the corresponding size of it to the ipc_size, it will be the best. Imagine that we could have scenario that with no UUID but the new added extended data, for this case, we should transmit the extended data to FW, no? 

So here I just wanted to make it simplest for us, once we have one extended data member, let's add all defined extended data to the ipc_size, though there could be some members of the extended data are 0s.

OTOH, If we are all aligned that having the UUID is the precondition of adding any new added extended data, and once it has UUID we will assume it has all other members alongside, then checking the nil UUID only looks fine to me.

> 
> Hey we could even want different components to have different lengths of extended data because why not? This would of course require more code changes but at least this `memcmp` would not have to change.

Reply:
>  Imagine that we could have scenario that with no UUID but the new added extended data, for this case, we should transmit the extended data to FW, no?

What are the chances of future, non-UUID extended data starting with 16 (!) zeroes?

Maybe this is clearer then:

```C
/* special value not just for UUIDs */
uint8_t nul_ext_data[SOF_UUID_SIZE] = {0};
...
if (memcmp(&swidget->comp_ext, &nul_ext_data, SOF_UUID_SIZE))
```

Reply:
> > Imagine that we could have scenario that with no UUID but the new added extended data, for this case, we should transmit the extended data to FW, no?
> 
> What are the chances of future, non-UUID extended data starting with 16 (!) zeroes?

You need to understand that the extended data is assigned when tokens are parsed, if no related tokens are parsed from the tplg files then it keeps the initiated 0s.

> 
> Maybe this is clearer then:
> 
> ```c
> /* special value not just for UUIDs */
> uint8_t nul_ext_data_prefix[SOF_UUID_SIZE] = {0};
> ...
> if (memcmp(&swidget->comp_ext, &nul_ext_data_prefix, SOF_UUID_SIZE))
> ```

We are adding the whole extended data (total_size += sizeof(swidget->comp_ext)), so comparing the whole struct is more legit to me, no? Why you are suggesting comparing the SOF_UUID_SIZE but adding the full extended struct to the comp->hdr.size?

Reply:
> You need to understand that the extended data is assigned when tokens are parsed, if no related tokens are parsed from the tplg files then it keeps the initiated 0s.

I think we keep talking across each other. You're talking about the present whereas I'm only talking about the future.

Let me try a different way: suppose someone wants to add in a future some new, optional extended data that is 128 bytes big. Do we really need to allocate 128 zeroes on the kernel stack to test when this big extended data is being used? I think we don't, I think checking the first 16 bytes is more than enough. Maybe even the first 8 bytes is enough. Do you think we will need to check all 128 bytes and why? If yes then it nul_ext_data has to be moved away from the stack: https://elinux.org/Kernel_Small_Stacks

Reply:
@marc-hb I think we can argue for both sides but even if the ext_data is extended beyond just the UUID, the fact remains that it cannot be all 0s because UUID is already part of it. So keeping this check as is seems fine to me.

Reply:
> the fact remains that it cannot be all 0s because UUID is already part of it. 

OK so if we assume future extended data always has a UUID prefix then there is really no need to look at anything past the UUID (and blow up the stack) which is exactly what I'm suggesting.

Reply:
@marc-hb as I commented above, If we are all aligned that having the UUID is the precondition of adding any new added extended data, and once it has UUID we will assume it has all other members alongside, then checking the nil UUID only looks fine to me.

And in a word, anyway, this is not a big deal to me as the extended data has only UUID at the moment, let's compare the UUID only at the moment and leave changes to future if needed.